src/auth/Crypto.cc

```diff
@@ -35,32 +35,49 @@
 #include "common/debug.h"
 #include <errno.h>
 
-int get_random_bytes(char *buf, int len)
+// use getentropy() if available. it uses the same source of randomness
+// as /dev/urandom without the filesystem overhead
+#ifdef HAVE_GETENTROPY
+
+#include <unistd.h>
+
+CryptoRandom::CryptoRandom() : fd(0) {}
+CryptoRandom::~CryptoRandom() = default;
+
+void CryptoRandom::get_bytes(char *buf, int len)
 {
-  int fd = TEMP_FAILURE_RETRY(::open("/dev/urandom", O_RDONLY));
-  if (fd < 0)
-    return -errno;
-  int ret = safe_read_exact(fd, buf, len);
-  VOID_TEMP_FAILURE_RETRY(::close(fd));
-  return ret;
+  auto ret = TEMP_FAILURE_RETRY(::getentropy(buf, len));
+  if (ret < 0) {
+    throw std::system_error(errno, std::system_category());
+  }
 }
 
-static int get_random_bytes(int len, bufferlist& bl)
+#else // !HAVE_GETENTROPY
+
+// open /dev/urandom once on construction and reuse the fd for all reads
+CryptoRandom::CryptoRandom()
+  : fd(TEMP_FAILURE_RETRY(::open("/dev/urandom", O_RDONLY)))
 {
-  char buf[len];
-  get_random_bytes(buf, len);
-  bl.append(buf, len);
-  return 0;
+  if (fd < 0) {
+    throw std::system_error(errno, std::system_category());
+  }
 }
 
-uint64_t get_random(uint64_t min_val, uint64_t max_val)
+CryptoRandom::~CryptoRandom()
 {
-  uint64_t r;
-  get_random_bytes((char *)&r, sizeof(r));
-  r = min_val + r % (max_val - min_val + 1);
-  return r;
+  VOID_TEMP_FAILURE_RETRY(::close(fd));
 }
 
+void CryptoRandom::get_bytes(char *buf, int len)
+{
+  auto ret = safe_read_exact(fd, buf, len);
+  if (ret < 0) {
```

Comment:
Note that C++ lets you declare a variable in a branch, so you can do this in one step if you want:
if(auto ret = safe_read_exact(fd, buf, len) > 0) 
 ...